### ISSUE TITLE
Semantic validation of data types with "format"

### ISSUE THREAD

**Issue by lrichert** [remotestorage/remotestorage.js#1162]
Hi,
the docs are stating

> JSON Schema object defining the actual structure and formatting of your data

and

> reject the promise with detailed validation errors in case the data format doesn’t match

This implies (at least for me) support of semantic validations by the `format` keyword. Getting fulfilled promises all the time, it took me a while to find out the tv4 package doesn't support `format` out of the box and seemingly ignores it silently.

So I want to suggest to either note this behavior in the docs or include the [tv4-formats package](https://www.npmjs.com/package/tv4-formats), a tv4 plugin containing some common string format constraints - and document that anyway, since it covers only a subset of the JSON Schema validation draft and that could lead to swallowed errors once again.

Finally, thank you for this piece of software. It spares me the whole server-side backend implementation on my current project and greatly increases reliability, I think.

**Comment by lrichert** [comment]
With this little thingy in `baseclient.js`
```js
/**
 * Add custom format validators.
 *
 * See {@link https://github.com/geraintluff/tv4#addformatformat-validationfunction} for more info.
 *
 * @param {object} format - JS object with format strings as keys and validation functions as values.
 **/
addValidationFormat: function(format) {
  tv4.addFormat(format);
},
```
it's possible to hook own validators like this:
```js
var Dates = { name: 'dates', builder: function(privateClient, publicClient) {
  privateClient.declareType('route', {
    "type": "object",
    "properties": {
      "date": {
        "type": "string",
        "format": "date-time"
      }
    }
  });

  privateClient.addValidationFormat({
    'date-time': function(value) {
      var dateTimeRegexp = /^[0-9]{4,}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]+|)(?:[+-][0-9]{2}:?(?:[0-9]{2}|)|Z)$/;
      if (dateTimeRegexp.test(value)) return null;
      return "A valid ISO 8601 date/time string expected";
    }
  });

  return { exports: { /* ... */ } }
}};
```

Or with `tv4-formats` in place:
```js
// var formats = require('tv4-formats');
privateClient.addValidationFormat(formats);
```

Just a sample. I don't know if it's a good idea to introduce public API coupled so tightly to the underlying tv4.

**Comment by skddc** [comment]
Thanks for opening this issue and presenting a possible solution!

I agree that it might be overkill to add this as a new dependency for all users, and I also agree that overly tight coupling to tv4 is probably not a good idea. However, I really like your solution with the new `BaseClient` function, and that function could also do something different, in case we ever remove tv4.

I'm also wondering, if we shouldn't just add the most popular formats to the core library though. My first guess would be that `date-time`, `date` and `url` are by far the most widely used, and everything else could then be custom to whatever the module needs. What do you think?

**Comment by lrichert** [comment]
Thank you really much for your kind answer!
This would be fantastic. These could be included, I think. They adhere to the JSON Schema specification, too.
- `date-time`: see my post above, scraped from [tv4-formats](https://github.com/ikr/tv4-formats/blob/master/src/dateTimeRegExp.js)
- `date`: same thing shortened: `[0-9]{4,}-[0-9]{2}-[0-9]{2}`
- `uri`: There's a pattern provided with [RFC3986, Appendix B](https://tools.ietf.org/html/rfc3986#appendix-B), I haven't tested it extensively though.

I would have loved to throw `email` in, but in search for a suitable regex I just found countless discussions, [monstrous patterns](https://stackoverflow.com/a/1903368/2715241) (this one is even obsolete and far from complete) often looking disturbingly similar to Malbolge code and even blog posts from people like [Phil Haack](https://haacked.com/archive/2007/08/21/i-knew-how-to-validate-an-email-address-until-i.aspx/) and Jeff Atwood dealing with the difficulties of this topic.
For me, it was one of the many *Gimme five minutes! Well... wait.* moments.

**Comment by DougReeder** [comment]
I agree that date-time, date and url are key. An email format that matches
an unanchored non-space character, the @ sign and another non-space
character is sufficient, I believe.
